### ISSUE TITLE
Features and improvement suggestions

### ISSUE THREAD

**Issue by tilap** [onubo/react-native-logs#12]
Thanks for this logger. It was a missing master piece in the RN ecosystem.

Here are some improvements suggestions I would be happy to have. Idk if I will have time to do some pull request, so I let you my thought here:

- namespacing the loggers: it let us focus on some part of the stack/app
- namespacing children: same as above, but to look deeper.
- context: attaching a persistent context to a logger during it's lifecycle brings helpful information when debuging
- a firebase adapter
- a gcp adapter

Tbh, I often use this awesome logger for isomorphic web app: https://github.com/christophehurpeau/nightingale . It brings the features mentioned above and could be a great source of inspiration.

One more time: thanks for this great package :)

**Comment by alessandro-bottamedi** [comment]
Thanks for the compliments and for your suggestions, I am evaluating how to implement the future ver 3. For now the features that I want to implement are:

- log namespacing
- better arguments concatenation
- GraphQl adapter
- Firebase adapter

surely I will be inspired by the logger that you recommended for namespacing, thanks!
Any contribution with a pull request is welcome!

**Comment by ca057** [comment]
Hey @alessandro-bottamedi - having firebase as a transport option would be nice. Is this already in implementation or is this something I could help out with?

**Comment by alessandro-bottamedi** [comment]
I haven't been able to work on it much in the last period, I plan to spend more time next month, for now my priority is to improve the logs concatenation and namespacing. 
So if you have started to write a firebase transport it would be very useful ;)

**Comment by ca057** [comment]
Alright, I might be able to look into it the next days and will let you know before I start coding anything to align beforehand :)

**Comment by chmac** [comment]
Any update on namespacing? This is the big one for me. I'm also wondering if I could somehow hook into the `debug` package and dump its logs in here. I already use it, and it's widely used in other packages, so that might provide a "one size fits all" log strategy, which would also support namespacing... 🤔

**Comment by alessandro-bottamedi** [comment]
@chmac @tilap
Each logger is designed to be independent, it is an object that has inside all the necessary configuration. This is to be able to create multiple loggers with different levels and transports.

The purpose of having namespace logs is to be able to display only the logos of a specific part of the app.

Currently the only way to do this is to create different loggers for different parts of the app and set a different severity for each logger according to which part of the app we want to examine:

```javascript
var logA = logger.createLogger();
var logB = logger.createLogger();
var logC = logger.createLogger();

logA.setSeverity('debug');
logB.setSeverity('info');
logC.setSeverity('error');

logA.debug('will be printed');
logB.debug('will not be printed');
logC.debug('will not be printed');

logA.info('will be printed');
logA.info('will be printed');
logA.info('will not be printed');
```

Now there are other several ways to implement namespaces:

1. Maintain the current system and then create more loggers

2. outsource the configuration for all loggers and set a namespace for each logger:

```javascript
import { logger } from 'react-native-logs';

const config = {
  levels: {
    trace: 0,
    info: 1,
    silly: 2,
    error: 3,
    mad: 4,
  },
};

logger.config(config);

var logA = logger.create('namespaceA');
var logB = logger.create('namespaceB');
var logC = logger.create('namespaceC');

logger.setSeverity('debug');
logger.setVisibleNamespace(['namespaceA','namespaceB']);

logA.info('will be printed');
logB.info('will be printed');
logC.info('will not be printed');

```

3. Add a `namespace` parameter to each log and add a `setVisibleNamespace` method to each logger objects:

```javascript
import { logger } from 'react-native-logs';

var log = logger.createLogger();

log.setVisibleNamespace(['namespaceA']);

log.debug('will be printed','namespaceA');
log.debug('will not be printed','namespaceB');

```

Which of these ways seems the best to you (1, 2, or 3)? 
others idea?
Thanks

**Comment by chmac** [comment]
@alessandro-bottamedi I personally would suggest following `debug`, which works like `debug.extend('namespace')`. They can be nested like so:

```typescript
const log = debug.extend('package')
const fileLogger = log.extend('file')
const fnLogger = fileLogger.extend('fn')
```

I ended up writing my own logger as I wanted to customise the presentation in the log files, etc. I haven't added filtering by severity support yet, but I've already added namespaces. I didn't want to litter my codebase with `log()` calls only to have to go back later and add contexts. To me this is the key feature, being able to tell which part of the code my logs come from. You can see what I've done here:

https://github.com/GenerousLabs/share/blob/main/expo/src/services/log/log.service.ts

The basic premise is the same as `debug`, it works with a `.extend()` method that appends a new namespace onto the logger. Then the log output is like `namespace1:namespace2:namespace3` etc.

**Comment by alessandro-bottamedi** [comment]
I've released the new v 3.0.0 with most of above suggestion, i close this issue, open another one issue for future suggestions...

Feedback on the v 3.0.0 are walcome! thanks again!
